app/src/main/java/io/particle/android/sdk/ui/DeviceListFragment.java

```diff
@@ -303,19 +309,23 @@ public boolean onBackPressed() {
     }
 
     private void addPhotonDevice() {
-        ParticleDeviceSetupLibrary.startDeviceSetup(getActivity());
+        ParticleDeviceSetupLibrary.startDeviceSetup(getActivity(), DeviceListActivity.class);
     }
 
     private void addSparkCoreDevice() {
-        String coreAppPkg = "io.spark.core.android";
-        // Is the spark core app already installed?
-        Intent intent = getActivity().getPackageManager().getLaunchIntentForPackage(coreAppPkg);
-        if (intent == null) {
-            // Nope.  Send the user to the store.
-            intent = new Intent(Intent.ACTION_VIEW)
-                    .setData(Uri.parse("market://details?id=" + coreAppPkg));
+        try {
+            String coreAppPkg = "io.spark.core.android";
+            // Is the spark core app already installed?
+            Intent intent = getActivity().getPackageManager().getLaunchIntentForPackage(coreAppPkg);
+            if (intent == null) {
+                // Nope.  Send the user to the store.
+                intent = new Intent(Intent.ACTION_VIEW)
+                        .setData(Uri.parse("market://details?id=" + coreAppPkg));
+            }
+            startActivity(intent);
+        } catch (ActivityNotFoundException ignored) {
+            Toast.makeText(getActivity(), "Cannot find spark core application.", Toast.LENGTH_SHORT).show();
```

Comment:
Whaaaaaaaat?  It's `startActivity()` throwing that exception, right?

If that intent wasn't `null`, the Spark Core app was found.  So... sometimes the play store can't be found?  Were we seeing this problem in China or something?

Reply:
Yep, most likely China, or devices without play store :P